### ISSUE TITLE
datagen: Unrecognized header

### ISSUE THREAD

**Issue by itoffshore** [lz4/lz4#385]
In [Alpine Linux](http://alpinelinux.org/) now we run the included unit tests for all packages. The test log shows all tests passing successfully but I see the following errors (possibly related to Alpine using Musl C not Glibc):

```
Error 44 : Unrecognized header : file cannot be decoded 
! ./datagen | ../programs/lz4 -tf

Error 44 : Unrecognized header : file cannot be decoded 
\n ---- pass-through mode ----
! ./datagen | ../programs/lz4 -d  > /dev/null

Error 44 : Unrecognized header : file cannot be decoded 
./datagen | ../programs/lz4 -df > /dev/null
```

& also in the decompression function tests `tmp` is not found (I think this may be due to different options between `coreutils` & the builtin `busybox` tools):
```
Decompression functions : 
Read : 3068 MB   ==> 0.39%   
Compressed 3221225472 bytes into 12676935 bytes ==> 0.39%                      
Completed in 19.69 sec 
stdin                : decoded 3221225472 bytes                                
  899    1-LZ4_decompress_fast           :     18092 ->   109.9 MB/s
  963    3-LZ4_decompress_fast_usingDict :     18092 ->   119.5 MB/s
 6323    4-LZ4_decompress_safe           :     18092 ->   111.9 MB/s
 6621    6-LZ4_decompress_safe_usingDict :     18092 ->   118.4 MB/s
 6923    7-LZ4_decompress_safe_partial   :     18092 ->   112.8 MB/s
 1175    8-LZ4_decompress_safe_forceExtD :     18092 ->   119.2 MB/s
 1206   ls: tmp: No such file or directory
make[1]: *** [Makefile:185: test-lz4-contentSize] Error 1
```

This is the test log in full:

```
./fasttest
./frametest -T3mn
Starting lz4frame tester (64-bits, 1.8.0)
Seed = 5207
gcc -Os -fomit-frame-pointer -g -Wall -Wextra -Wundef -Wcast-qual -Wcast-align -Wshadow -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wpointer-arith -Wstrict-aliasing=1  -I../lib -I../programs -DXXH_NAMESPACE=LZ4_ -I../lib -DXXH_NAMESPACE=LZ4_  -c -o bench.o bench.c
Basic tests completed 
    0   gcc -Os -fomit-frame-pointer -g -Wall -Wextra -Wundef -Wcast-qual -Wcast-align -Wshadow -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wpointer-arith -Wstrict-aliasing=1  -I../lib -I../programs -DXXH_NAMESPACE=LZ4_ -I../lib -DXXH_NAMESPACE=LZ4_  -c -o datagen.o datagen.c
gcc -Os -fomit-frame-pointer -g -Wall -Wextra -Wundef -Wcast-qual -Wcast-align -Wshadow -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wpointer-arith -Wstrict-aliasing=1  -I../lib -I../programs -DXXH_NAMESPACE=LZ4_ -I../lib -DXXH_NAMESPACE=LZ4_ -Wl,--as-needed ../lib/lz4frame.o ../lib/xxhash.o ../lib/lz4hc.o ../lib/lz4.o bench.o lz4cli.o lz4io.o datagen.o -o lz4
ln -s lz4 unlz4
make[2]: Leaving directory '/home/stuart/aports/main/lz4/src/lz4-1.8.0/programs'
./fullbench --no-prompt -i1 COPYING
gcc -Os -fomit-frame-pointer -g -Wall -Wextra -Wundef -Wcast-qual -Wcast-align -Wshadow -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wpointer-arith -Wstrict-aliasing=1  -I../lib -I../programs -DXXH_NAMESPACE=LZ4_ -I../lib -DXXH_NAMESPACE=LZ4_  -c -o bench.o bench.c
*** LZ4 speed analyzer v1.8.0 64-bits, by Yann Collet ***
- 1 iterations -
 COPYING :                                                                     
Compression functions : 
gcc -Os -fomit-frame-pointer -g -Wall -Wextra -Wundef -Wcast-qual -Wcast-align -Wshadow -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wpointer-arith -Wstrict-aliasing=1  -I../lib -I../programs -DXXH_NAMESPACE=LZ4_ -I../lib -DXXH_NAMESPACE=LZ4_  -c -o datagen.o datagen.c
   15   gcc -Os -fomit-frame-pointer -g -Wall -Wextra -Wundef -Wcast-qual -Wcast-align -Wshadow -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wpointer-arith -Wstrict-aliasing=1  -I../lib -I../programs -DXXH_NAMESPACE=LZ4_ -I../lib -DXXH_NAMESPACE=LZ4_ -Wl,--as-needed ../lib/lz4frame.o ../lib/xxhash.o ../lib/lz4hc.o ../lib/lz4.o bench.o lz4cli.o lz4io.o datagen.o -o lz4
   56   gcc -Os -fomit-frame-pointer -g -Wall -Wextra -Wundef -Wcast-qual -Wcast-align -Wshadow -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wpointer-arith -Wstrict-aliasing=1  -I../lib -I../programs -DXXH_NAMESPACE=LZ4_ -I../lib -DXXH_NAMESPACE=LZ4_ -Wl,--as-needed ../lib/lz4frame.o ../lib/xxhash.o ../lib/lz4hc.o ../lib/lz4.o bench.o lz4cli.o lz4io.o datagen.o -o lz4
ln -s lz4 lz4cat
make[2]: Leaving directory '/home/stuart/aports/main/lz4/src/lz4-1.8.0/programs'
gcc -Os -fomit-frame-pointer -g -Wall -Wextra -Wundef -Wcast-qual -Wcast-align -Wshadow -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wpointer-arith -Wstrict-aliasing=1  -I../lib -I../programs -DXXH_NAMESPACE=LZ4_ -I../lib -DXXH_NAMESPACE=LZ4_  -c -o lz4io.o lz4io.c
ln -s lz4 lz4c
make[2]: Leaving directory '/home/stuart/aports/main/lz4/src/lz4-1.8.0/programs'
\n ---- test lz4c version ----
./datagen -g256MB | ../programs/lz4c -l -v    | ../programs/lz4c   -t
*** LZ4 command line interface 64-bits v1.8.0, by Yann Collet ***
! Generating compressed LZ4 using Legacy format (deprecated) ! 
   73   ./fuzzer -T3mn
Starting LZ4 fuzzer (64-bits, v1.8.0)
Seed = 9151
Overflow tests : 64 bits mode : no overflow 
  106   All unit tests completed successfully compressionLevel=9 
Overflow tests : 64 bits mode : no overflow 
  184   All unit tests completed successfully compressionLevel=11 
    0   gcc -Os -fomit-frame-pointer -g -Wall -Wextra -Wundef -Wcast-qual -Wcast-align -Wshadow -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wpointer-arith -Wstrict-aliasing=1  -I../lib -I../programs -DXXH_NAMESPACE=LZ4_ -I../lib -DXXH_NAMESPACE=LZ4_ -Wl,--as-needed ../lib/lz4frame.o ../lib/xxhash.o ../lib/lz4hc.o ../lib/lz4.o bench.o lz4cli.o lz4io.o datagen.o -o lz4
make[2]: Leaving directory '/home/stuart/aports/main/lz4/src/lz4-1.8.0/programs'
\n ---- test opt-parser ----
./datagen -g16KB      | ../programs/lz4 -12      | ../programs/lz4 -t
\n ---- test lz4 basic compression/decompression ----
./datagen -g0     | ../programs/lz4 -v     | ../programs/lz4 -t

stdin                : decoded 16384 bytes                                     
./datagen -P10        | ../programs/lz4 -12B4    | ../programs/lz4 -t

using blocks of size 64 KB 
*** LZ4 command line interface 64-bits v1.8.0, by Yann Collet ***
                                                                               
Compressed 0 bytes into 15 bytes ==> 1500.00%
stdin                : decoded 0 bytes                                         
./datagen -g16KB  | ../programs/lz4 -9     | ../programs/lz4 -t
stdin                : decoded 65536 bytes                                     
./datagen -g256K      | ../programs/lz4 -12B4D   | ../programs/lz4 -t
using blocks of size 64 KB 

stdin                : decoded 16384 bytes                                     
./datagen -g20KB > tmpSrc

../programs/lz4 < tmpSrc   | ../programs/lz4 -d > tmpRes
diff -q tmpSrc tmpRes
../programs/lz4 --no-frame-crc < tmpSrc | ../programs/lz4 -d > tmpRes
diff -q tmpSrc tmpRes
./datagen         | ../programs/lz4        | ../programs/lz4 -t

stdin                : decoded 65536 bytes                                     
./datagen -g6M -P99 | ../programs/lz4 -9BD | ../programs/lz4 -t

stdin                : decoded 262144 bytes                                    
./datagen -g512K -P25 | ../programs/lz4 -12BD    | ../programs/lz4 -t
    8   

stdin                : decoded 524288 bytes                                    
./datagen -g1M        | ../programs/lz4 -12B5    | ../programs/lz4 -t
using blocks of size 256 KB 
stdin                : decoded 6291456 bytes                                   
./datagen -g17M   | ../programs/lz4 -9v    | ../programs/lz4 -qt
*** LZ4 command line interface 64-bits v1.8.0, by Yann Collet ***
    9   
stdin                : decoded 1048576 bytes                                   
./datagen -g2M -P99   | ../programs/lz4 -11B4D   | ../programs/lz4 -t
  327   using blocks of size 64 KB 
  383   
stdin                : decoded 2097152 bytes                                   
./datagen -g4M        | ../programs/lz4 -11vq    | ../programs/lz4 -qt

  393    1-LZ4_compress_default         :    18092 ->    10582 (58.49%),  196.6 MB/s
  412   
./datagen -g8M        | ../programs/lz4 -11B4    | ../programs/lz4 -t
   42   using blocks of size 64 KB 
Compressed 17825792 bytes into 8086251 bytes ==> 45.36%                        
./datagen -g33M   | ../programs/lz4 --no-frame-crc | ../programs/lz4 -t
   45   
stdin                : decoded 34603008 bytes                                  
./datagen -g256MB | ../programs/lz4 -vqB4D | ../programs/lz4 -t
  526   using blocks of size 64 KB 
   61   
Compressed 268435456 bytes into 156514661 bytes ==> 58.31%                     
stdin                : decoded 268435456 bytes                                 
\n ---- test multiple files ----
../programs/lz4 -f -m tmp*
  549   ls -ls tmp*
    64 -rw-r--r--    1 stuart   stuart       65536 Aug 20 16:04 tmp1
    36 -rw-r--r--    1 stuart   stuart       33368 Aug 20 16:04 tmp1.lz4
   100 -rw-r--r--    1 stuart   stuart      102400 Aug 20 16:04 tmp2
    56 -rw-r--r--    1 stuart   stuart       56193 Aug 20 16:04 tmp2.lz4
  1024 -rw-r--r--    1 stuart   stuart     1048576 Aug 20 16:04 tmp3
   596 -rw-r--r--    1 stuart   stuart      606770 Aug 20 16:04 tmp3.lz4
    20 -rw-r--r--    1 stuart   stuart       20480 Aug 20 16:04 tmpRes
    16 -rw-r--r--    1 stuart   stuart       12461 Aug 20 16:04 tmpRes.lz4
    20 -rw-r--r--    1 stuart   stuart       20480 Aug 20 16:04 tmpSrc
    16 -rw-r--r--    1 stuart   stuart       12461 Aug 20 16:04 tmpSrc.lz4
../programs/lz4 -df -m *.lz4
ls -ls tmp*
    64 -rw-r--r--    1 stuart   stuart       65536 Aug 20 16:04 tmp1
    36 -rw-r--r--    1 stuart   stuart       33368 Aug 20 16:04 tmp1.lz4
   100 -rw-r--r--    1 stuart   stuart      102400 Aug 20 16:04 tmp2
    56 -rw-r--r--    1 stuart   stuart       56193 Aug 20 16:04 tmp2.lz4
  1024 -rw-r--r--    1 stuart   stuart     1048576 Aug 20 16:04 tmp3
   596 -rw-r--r--    1 stuart   stuart      606770 Aug 20 16:04 tmp3.lz4
    20 -rw-r--r--    1 stuart   stuart       20480 Aug 20 16:04 tmpRes
    16 -rw-r--r--    1 stuart   stuart       12461 Aug 20 16:04 tmpRes.lz4
    20 -rw-r--r--    1 stuart   stuart       20480 Aug 20 16:04 tmpSrc
    16 -rw-r--r--    1 stuart   stuart       12461 Aug 20 16:04 tmpSrc.lz4
../programs/lz4 -f -m tmp1 notHere tmp2; echo $?
notHere: No such file or directory 
1
\n ---- test sparse file support ----
./datagen -g5M  -P100 > tmpSrc
   62   
../programs/lz4 -B4D tmpSrc | ../programs/lz4 -dv --sparse > tmpB4
*** LZ4 command line interface 64-bits v1.8.0, by Yann Collet ***
using blocks of size 64 KB 
stdin                : decoded 5242880 bytes                                   
diff -s tmpSrc tmpB4
Files tmpSrc and tmpB4 are identical
../programs/lz4 -B5D tmpSrc | ../programs/lz4 -dv --sparse > tmpB5
using blocks of size 256 KB 
*** LZ4 command line interface 64-bits v1.8.0, by Yann Collet ***

stdin                : decoded 8388608 bytes                                   
./datagen -g16M -P90  | ../programs/lz4 -11B5    | ../programs/lz4 -t
using blocks of size 256 KB 
stdin                : decoded 5242880 bytes                                   
diff -s tmpSrc tmpB5
Files tmpSrc and tmpB5 are identical
../programs/lz4 -B6D tmpSrc | ../programs/lz4 -dv --sparse > tmpB6
*** LZ4 command line interface 64-bits v1.8.0, by Yann Collet ***
using blocks of size 1024 KB 
stdin                : decoded 5242880 bytes                                   
diff -s tmpSrc tmpB6
   70   Files tmpSrc and tmpB6 are identical
../programs/lz4 -B7D tmpSrc | ../programs/lz4 -dv --sparse > tmpB7
*** LZ4 command line interface 64-bits v1.8.0, by Yann Collet ***
using blocks of size 4096 KB 
stdin                : decoded 5242880 bytes                                   
diff -s tmpSrc tmpB7
Files tmpSrc and tmpB7 are identical
../programs/lz4 tmpSrc | ../programs/lz4 -dv --no-sparse > tmpNoSparse
*** LZ4 command line interface 64-bits v1.8.0, by Yann Collet ***
stdin                : decoded 5242880 bytes                                   
diff -s tmpSrc tmpNoSparse
Files tmpSrc and tmpNoSparse are identical
ls -ls tmp*
  5120 -rw-r--r--    1 stuart   stuart     5242880 Aug 20 16:04 tmpB4
  5120 -rw-r--r--    1 stuart   stuart     5242880 Aug 20 16:04 tmpB5
  5120 -rw-r--r--    1 stuart   stuart     5242880 Aug 20 16:04 tmpB6
  5120 -rw-r--r--    1 stuart   stuart     5242880 Aug 20 16:04 tmpB7
  5120 -rw-r--r--    1 stuart   stuart     5242880 Aug 20 16:04 tmpNoSparse
  5120 -rw-r--r--    1 stuart   stuart     5242880 Aug 20 16:04 tmpSrc
./datagen -s1 -g1200007 -P100 | ../programs/lz4 | ../programs/lz4 -dv --sparse > tmpOdd   # Odd size file (to generate non-full last block)
  738   *** LZ4 command line interface 64-bits v1.8.0, by Yann Collet ***

stdin                : decoded 1200007 bytes                                   
./datagen -s1 -g1200007 -P100 | diff -s - tmpOdd

Files - and tmpOdd are identical
ls -ls tmpOdd
  1172 -rw-r--r--    1 stuart   stuart     1200007 Aug 20 16:04 tmpOdd
\n Compatibility with Console :
echo "Hello World 1 !" | ../programs/lz4 | ../programs/lz4 -d -c
Hello World 1 !
echo "Hello World 2 !" | ../programs/lz4 | ../programs/lz4 -d | cat
Hello World 2 !
echo "Hello World 3 !" | ../programs/lz4 --no-frame-crc | ../programs/lz4 -d -c
Hello World 3 !
\n Compatibility with Append :
./datagen -P100 -g1M > tmp1M

cat tmp1M tmp1M > tmp2M
../programs/lz4 -B5 -v tmp1M tmpC
using blocks of size 256 KB 
*** LZ4 command line interface 64-bits v1.8.0, by Yann Collet ***
Compressed 1048576 bytes into 4192 bytes ==> 0.40%                             
../programs/lz4 -d -v tmpC tmpR
*** LZ4 command line interface 64-bits v1.8.0, by Yann Collet ***
tmpC                 : decoded 1048576 bytes                                   
../programs/lz4 -d -v tmpC >> tmpR
*** LZ4 command line interface 64-bits v1.8.0, by Yann Collet ***
tmpC                 : decoded 1048576 bytes                                   
ls -ls tmp*
  1024 -rw-r--r--    1 stuart   stuart     1048576 Aug 20 16:04 tmp1M
  2048 -rw-r--r--    1 stuart   stuart     2097152 Aug 20 16:04 tmp2M
     8 -rw-r--r--    1 stuart   stuart        4192 Aug 20 16:04 tmpC
  1032 -rw-r--r--    1 stuart   stuart     2097152 Aug 20 16:04 tmpR
diff tmp2M tmpR
\n ---- test frame concatenation ----
cat nonempty.test empty.test nonempty.test > orig.test
cat nonempty.lz4.test empty.lz4.test nonempty.lz4.test > concat.lz4.test
../programs/lz4 -d concat.lz4.test > result.test
sdiff orig.test result.test
hi								hi
hi								hi
frame concatenation test completed
\n ---- bench mode ----
../programs/lz4 -bi1
  778    2-LZ4_compress_default(small d :    18092 ->    10582 (58.49%),  188.6 MB/s
  877   
stdin                : decoded 16777216 bytes                                  
./datagen -g32M -P10  | ../programs/lz4 -11B5D   | ../programs/lz4 -t
using blocks of size 256 KB 
  127    3-LZ4_compress_fast(0)         :    18092 ->    10582 (58.49%),  194.7 MB/s
 1#77    =-Synthetic 50%     :  10000000 ->   5859357 (1.707), 169.8 MB/s , 366.3 MB/s
\n ---- test mode ----
! ./datagen | ../programs/lz4 -t

Error 44 : Unrecognized header : file cannot be decoded 
! ./datagen | ../programs/lz4 -tf

Error 44 : Unrecognized header : file cannot be decoded 
\n ---- pass-through mode ----
! ./datagen | ../programs/lz4 -d  > /dev/null

Error 44 : Unrecognized header : file cannot be decoded 
./datagen | ../programs/lz4 -df > /dev/null

../programs/lz4 -dcf tmp1
Hello World !
../programs/lz4 -dcfm tmp1 tmp2
Hello World !
from underground...
\n ---- test cli ----
! ../programs/lz4     file-does-not-exist
Compressed filename will be : file-does-not-exist.lz4 
file-does-not-exist: No such file or directory 
! ../programs/lz4 -f  file-does-not-exist
Compressed filename will be : file-does-not-exist.lz4 
file-does-not-exist: No such file or directory 
! ../programs/lz4 -fm file1-dne file2-dne
file1-dne: No such file or directory 
file2-dne: No such file or directory 
! ../programs/lz4 -fm file1-dne file2-dne
file1-dne: No such file or directory 
file2-dne: No such file or directory 
\n ---- test original size support ----
./datagen -g15M > tmp
 1227   
../programs/lz4 -v tmp | ../programs/lz4 -t
*** LZ4 command line interface 64-bits v1.8.0, by Yann Collet ***
Compressed 15728640 bytes into 9227754 bytes ==> 58.67%                        
stdin                : decoded 15728640 bytes                                  
../programs/lz4 -v --content-size tmp | ../programs/lz4 -d > tmp2
*** LZ4 command line interface 64-bits v1.8.0, by Yann Collet ***
Compressed 15728640 bytes into 9227762 bytes ==> 58.67%                        
diff -s tmp tmp2
  150   Files tmp and tmp2 are identical
# test large size [2-4] GB
*** LZ4 command line interface 64-bits v1.8.0, by Yann Collet ***
_POSIX_C_SOURCE defined: 200112L
_POSIX_VERSION defined: 200809L
PLATFORM_POSIX_VERSION defined: 200809L
Using stdin for input
Using stdout for output
Sparse File Support is automatically disabled on stdout ; try --sparse 
Read : 76 MB   ==> 0.39%   
stdin                : decoded 33554432 bytes                                  
\n ---- test huge files compression/decompression ----
./datagen -g6GB   | ../programs/lz4 -vB5D  | ../programs/lz4 -qt
using blocks of size 256 KB 
*** LZ4 command line interface 64-bits v1.8.0, by Yann Collet ***
Read : 88 MB   ==> 0.39%    4-LZ4_compress_fast(1)         :    18092 ->    10582 (58.49%),  187.6 MB/s
Read : 120 MB   ==> 0.39%   
stdin                : decoded 268435456 bytes                                 
Read : 124 MB   ==> 0.39%   ../programs/lz4 --rm -f tmp tmp.lz4
Compressed 12 bytes into 31 bytes ==> 258.33%                                  
test ! -f tmp                      # must fail (--rm)
Read : 128 MB   ==> 0.39%   test   -f tmp.lz4
../programs/lz4cat tmp.lz4           # must display hello world
Read : 132 MB   ==> 0.39%   hello world
test   -f tmp.lz4
../programs/unlz4 --rm tmp.lz4 tmp
tmp.lz4              : decoded 12 bytes                                        
test   -f tmp
test ! -f tmp.lz4                  # must fail (--rm)
Read : 140 MB   ==> 0.39%   test ! -f tmp.lz4.lz4              # must fail (unlz4)
../programs/lz4cat tmp               # pass-through mode
hello world
test   -f tmp
Read : 144 MB   ==> 0.39%   test ! -f tmp.lz4                  # must fail (lz4cat)
Read : 148 MB   ==> 0.39%   ../programs/lz4 tmp tmp.lz4                 # creates tmp.lz4
Compressed 12 bytes into 31 bytes ==> 258.33%                                  
../programs/lz4cat < tmp.lz4 > tmp3  # checks lz4cat works with stdin (#285)
diff -q tmp tmp3
../programs/lz4cat < tmp > tmp2      # checks lz4cat works with stdin (#285)
diff -q tmp tmp2
cp tmp ./-d
../programs/lz4 --rm -- -d -d.lz4           # compresses ./d into ./-d.lz4
Compressed 12 bytes into 31 bytes ==> 258.33%                                  
test   -f ./-d.lz4
test ! -f ./-d
mv ./-d.lz4 ./-z
../programs/lz4 -d --rm -- -z tmp4          # uncompresses ./-z into tmp4
-z                   : decoded 12 bytes                                        
test ! -f ./-z
diff -q tmp tmp4
../programs/lz4 -f tmp
Compressed filename will be : tmp.lz4 
Compressed 12 bytes into 31 bytes ==> 258.33%                                  
cat tmp >> tmp.lz4
../programs/lz4 -f tmp.lz4         # uncompress valid frame followed by invalid data
Decoding file tmp 
Stream followed by undecodable data at position 35 
tmp.lz4              : decoded 12 bytes                                        
../programs/lz4 -BX tmp -c -q | ../programs/lz4 -tv  # test block checksum
*** LZ4 command line interface 64-bits v1.8.0, by Yann Collet ***
stdin                : decoded 12 bytes                                        
Read : 328 MB   ==> 0.39%    5-LZ4_compress_fast(2)         :    18092 ->    10951 (60.53%),  193.2 MB/s
  339    6-LZ4_compress_fast(17)        :    18092 ->    15583 (86.13%),  261.6 MB/s
Read : 812 MB   ==> 0.39%    7-LZ4_compress_fast_extState(0 :    18092 ->    10582 (58.49%),  180.8 MB/s
Read : 1116 MB   ==> 0.39%    8-LZ4_compress_fast_continue(0 :    18092 ->    10647 (58.85%),  163.5 MB/s
Read : 1420 MB   ==> 0.39%   10-LZ4_compress_HC              :    18092 ->     9050 (50.02%),   26.1 MB/s
Read : 1680 MB   ==> 0.39%   12-LZ4_compress_HC_extStateHC   :    18092 ->     9050 (50.02%),   36.3 MB/s
Read : 1976 MB   ==> 0.39%   14-LZ4_compress_HC_continue     :    18092 ->     9050 (50.02%),   36.4 MB/s
Read : 2248 MB   ==> 0.39%   20-LZ4_compress_forceDict       :    18092 ->    10647 (58.85%),  164.8 MB/s
Read : 2484 MB   ==> 0.39%   30-LZ4F_compressFrame           :    18092 ->    10597 (58.57%),  181.1 MB/s
Read : 2704 MB   ==> 0.39%   40-LZ4_saveDict                 :    18092 ->    18092 (100.0%),20048.8 MB/s
 5144   41-LZ4_saveDictHC               :    18092 ->    18092 (100.0%),19736.6 MB/s
Decompression functions : 
Read : 3068 MB   ==> 0.39%   
Compressed 3221225472 bytes into 12676935 bytes ==> 0.39%                      
Completed in 19.69 sec 
stdin                : decoded 3221225472 bytes                                
  899    1-LZ4_decompress_fast           :     18092 ->   109.9 MB/s
  963    3-LZ4_decompress_fast_usingDict :     18092 ->   119.5 MB/s
 6323    4-LZ4_decompress_safe           :     18092 ->   111.9 MB/s
 6621    6-LZ4_decompress_safe_usingDict :     18092 ->   118.4 MB/s
 6923    7-LZ4_decompress_safe_partial   :     18092 ->   112.8 MB/s
 1175    8-LZ4_decompress_safe_forceExtD :     18092 ->   119.2 MB/s
 1206   ls: tmp: No such file or directory
make[1]: *** [Makefile:185: test-lz4-contentSize] Error 1
make[1]: *** Waiting for unfinished jobs....
 1256    9-LZ4F_decompress               :     18092 ->   114.3 MB/s
20081   
Compressed 6442450944 bytes into 3750754317 bytes ==> 58.22%                   
./datagen -g6GB   | ../programs/lz4 -v5BD  | ../programs/lz4 -qt
*** LZ4 command line interface 64-bits v1.8.0, by Yann Collet ***
All tests completed   
   4801 /   4801   - all tests completed successfully 
compression ratio: 55.071%
HC compression ratio: 51.888%
ratio with dict: 35.480%

Compressed 6442450944 bytes into 2896075899 bytes ==> 44.95%                   
make[1]: Leaving directory '/home/stuart/aports/main/lz4/src/lz4-1.8.0/tests'
make: *** [Makefile:127: test] Error 2
```
__________________________________________________
& this is the shorter test log from installing `coreutils` into the build environment:
__________________________________________________

```
\n ---- test opt-parser ----
./datagen -g16KB      | ../programs/lz4 -12      | ../programs/lz4 -t

stdin                : decoded 16384 bytes                                     
./datagen -P10        | ../programs/lz4 -12B4    | ../programs/lz4 -t
using blocks of size 64 KB 

stdin                : decoded 65536 bytes                                     
./datagen -g256K      | ../programs/lz4 -12B4D   | ../programs/lz4 -t
using blocks of size 64 KB 

gcc -Os -fomit-frame-pointer -g -Wall -Wextra -Wundef -Wcast-qual -Wcast-align -Wshadow -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wpointer-arith -Wstrict-aliasing=1  -I../lib -I../programs -DXXH_NAMESPACE=LZ4_ -I../lib -DXXH_NAMESPACE=LZ4_  -c -o bench.o bench.c
\n ---- test multiple files ----
../programs/lz4 -f -m tmp*
stdin                : decoded 262144 bytes                                    
./datagen -g512K -P25 | ../programs/lz4 -12BD    | ../programs/lz4 -t
\n ---- test sparse file support ----
./datagen -g5M  -P100 > tmpSrc

ls -ls tmp*

../programs/lz4 -B4D tmpSrc | ../programs/lz4 -dv --sparse > tmpB4
  64 -rw-r--r-- 1 stuart stuart   65536 Aug 20 17:57 tmp1*** LZ4 command line interface 64-bits v1.8.0, by Yann Collet ***

using blocks of size 64 KB 
  36 -rw-r--r-- 1 stuart stuart   33368 Aug 20 17:57 tmp1.lz4
 100 -rw-r--r-- 1 stuart stuart  102400 Aug 20 17:57 tmp2
  56 -rw-r--r-- 1 stuart stuart   56193 Aug 20 17:57 tmp2.lz4
1024 -rw-r--r-- 1 stuart stuart 1048576 Aug 20 17:57 tmp3
 596 -rw-r--r-- 1 stuart stuart  606770 Aug 20 17:57 tmp3.lz4
5120 -rw-r--r-- 1 stuart stuart 5242880 Aug 20 17:57 tmpSrc
../programs/lz4 -df -m *.lz4
ls -ls tmp*
  64 -rw-r--r-- 1 stuart stuart   65536 Aug 20 17:57 tmp1
  36 -rw-r--r-- 1 stuart stuart   33368 Aug 20 17:57 tmp1.lz4
 100 -rw-r--r-- 1 stuart stuart  102400 Aug 20 17:57 tmp2
  56 -rw-r--r-- 1 stuart stuart   56193 Aug 20 17:57 tmp2.lz4
1024 -rw-r--r-- 1 stuart stuart 1048576 Aug 20 17:57 tmp3
 596 -rw-r--r-- 1 stuart stuart  606770 Aug 20 17:57 tmp3.lz4
 704 -rw-r--r-- 1 stuart stuart  720896 Aug 20 17:57 tmpB4
5120 -rw-r--r-- 1 stuart stuart 5242880 Aug 20 17:57 tmpSrc
../programs/lz4 -f -m tmp1 notHere tmp2; echo $?
notHere: No such file or directory 
1
\n ---- test frame concatenation ----
cat nonempty.test empty.test nonempty.test > orig.test
cat nonempty.lz4.test empty.lz4.test nonempty.lz4.test > concat.lz4.test
../programs/lz4 -d concat.lz4.test > result.test
sdiff orig.test result.test
hi								hi
hi								hi
frame concatenation test completed
\n ---- bench mode ----
../programs/lz4 -bi1
gcc -Os -fomit-frame-pointer -g -Wall -Wextra -Wundef -Wcast-qual -Wcast-align -Wshadow -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wpointer-arith -Wstrict-aliasing=1  -I../lib -I../programs -DXXH_NAMESPACE=LZ4_ -I../lib -DXXH_NAMESPACE=LZ4_ -Wl,--as-needed ../lib/lz4frame.o ../lib/xxhash.o ../lib/lz4hc.o ../lib/lz4.o bench.o lz4cli.o lz4io.o datagen.o -o lz4
stdin                : decoded 5242880 bytes                                   
diff -s tmpSrc tmpB4
diff: tmpSrc: No such file or directory
diff: tmpB4: No such file or directory
make[1]: *** [Makefile:149: test-lz4-sparse] Error 2
make[1]: *** Waiting for unfinished jobs....
gcc -Os -fomit-frame-pointer -g -Wall -Wextra -Wundef -Wcast-qual -Wcast-align -Wshadow -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wpointer-arith -Wstrict-aliasing=1  -I../lib -I../programs -DXXH_NAMESPACE=LZ4_ -I../lib -DXXH_NAMESPACE=LZ4_  -c -o lz4cli.o lz4cli.c
ln -s lz4 lz4cat
make[2]: Leaving directory '/home/stuart/aports/main/lz4/src/lz4-1.8.0/programs'
stdin                : decoded 524288 bytes                                    
./datagen -g1M        | ../programs/lz4 -12B5    | ../programs/lz4 -t
using blocks of size 256 KB 
gcc -Os -fomit-frame-pointer -g -Wall -Wextra -Wundef -Wcast-qual -Wcast-align -Wshadow -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wpointer-arith -Wstrict-aliasing=1  -I../lib -I../programs -DXXH_NAMESPACE=LZ4_ -I../lib -DXXH_NAMESPACE=LZ4_ -Wl,--as-needed ../lib/lz4frame.o ../lib/xxhash.o ../lib/lz4hc.o ../lib/lz4.o bench.o lz4cli.o lz4io.o datagen.o -o lz4
ln -s lz4 lz4c
make[2]: Leaving directory '/home/stuart/aports/main/lz4/src/lz4-1.8.0/programs'

stdin                : decoded 1048576 bytes                                   
./datagen -g2M -P99   | ../programs/lz4 -11B4D   | ../programs/lz4 -t
using blocks of size 64 KB 

stdin                : decoded 2097152 bytes                                   
./datagen -g4M        | ../programs/lz4 -11vq    | ../programs/lz4 -qt

./datagen -g8M        | ../programs/lz4 -11B4    | ../programs/lz4 -t
using blocks of size 64 KB 
gcc -Os -fomit-frame-pointer -g -Wall -Wextra -Wundef -Wcast-qual -Wcast-align -Wshadow -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wpointer-arith -Wstrict-aliasing=1  -I../lib -I../programs -DXXH_NAMESPACE=LZ4_ -I../lib -DXXH_NAMESPACE=LZ4_ -Wl,--as-needed ../lib/lz4frame.o ../lib/xxhash.o ../lib/lz4hc.o ../lib/lz4.o bench.o lz4cli.o lz4io.o datagen.o -o lz4
ln -s lz4 unlz4
make[2]: Leaving directory '/home/stuart/aports/main/lz4/src/lz4-1.8.0/programs'
 /-Synthetic 50%     :  10000000 ->   5859357 (1.707), 300.0 MB/s
stdin                : decoded 8388608 bytes                                   
./datagen -g16M -P90  | ../programs/lz4 -11B5    | ../programs/lz4 -t
using blocks of size 256 KB 

stdin                : decoded 16777216 bytes                                  
./datagen -g32M -P10  | ../programs/lz4 -11B5D   | ../programs/lz4 -t
using blocks of size 256 KB 
 \-Synthetic 50%     :  10000000 ->   5859357 (1.707), 316.8 MB/s , 405.0 MB/s
stdin                : decoded 33554432 bytes                                  
 1#Synthetic 50%     :  10000000 ->   5859357 (1.707), 316.8 MB/s , 432.8 MB/s
\n ---- test mode ----
! ./datagen | ../programs/lz4 -t

Error 44 : Unrecognized header : file cannot be decoded 
! ./datagen | ../programs/lz4 -tf

Error 44 : Unrecognized header : file cannot be decoded 
\n ---- pass-through mode ----
! ./datagen | ../programs/lz4 -d  > /dev/null

Error 44 : Unrecognized header : file cannot be decoded 
./datagen | ../programs/lz4 -df > /dev/null

../programs/lz4 -dcf tmp1
Hello World !
../programs/lz4 -dcfm tmp1 tmp2
Hello World !
from underground...
\n ---- test cli ----
! ../programs/lz4     file-does-not-exist
Compressed filename will be : file-does-not-exist.lz4 
file-does-not-exist: No such file or directory 
! ../programs/lz4 -f  file-does-not-exist
Compressed filename will be : file-does-not-exist.lz4 
file-does-not-exist: No such file or directory 
! ../programs/lz4 -fm file1-dne file2-dne
file1-dne: No such file or directory 
file2-dne: No such file or directory 
! ../programs/lz4 -fm file1-dne file2-dne
file1-dne: No such file or directory 
file2-dne: No such file or directory 
make[1]: Leaving directory '/home/stuart/aports/main/lz4/src/lz4-1.8.0/tests'
make: *** [Makefile:127: test] Error 2

```
The easiest way to duplicate this is to setup an [Alpine LXC container](https://wiki.alpinelinux.org/wiki/LXC#Alpine_Template) & `apk add alpine-sdk`

You will also need to `apk add diffutils` as the tests seem to use some options not in the `busybox` `diff` builtin command.

If I can help to resolve these problems please let me know.

**Comment by Cyan4973** [comment]
```
! ./datagen | ../programs/lz4 -t

Error 44 : Unrecognized header : file cannot be decoded 
```

The result of this test is correct, and it is an error.
We expect `lz4` to complain about an unsupported header when fed with noise, which is what `datagen` does.
The `!` at the beginning of the line is supposed to make an error become a success, and vice versa.

I see that `make` nonetheless fails later on, but it's unclear at this stage for which reason. It's not necessarily related to mentioned tests, since it fails later.

**Comment by mwgamera** [comment]
The tests can not be run in parallel because of how they use temporary files and abuild sets MAKEFLAGS to -j2 by default.

**Comment by itoffshore** [comment]
ok - this issue can be closed then

**Comment by Cyan4973** [comment]
For information, there is a PR which tries to make `make test` compatible with `-j#` : 
https://github.com/lz4/lz4/pull/386

**Comment by itoffshore** [comment]
yes I've used the diff in the `1.8.0` update https://github.com/alpinelinux/aports/pull/2161

* `-j4` works ok always
* `-j5` & `-j6` builds sometimes / sometimes not

I had to also `rm -f programs/lz4c` before `make install` (where it was created again)

**Comment by mwgamera** [comment]
Ok, this is more subtle and related to why many people consider recursive use of make evil.
An absolutely minimal fix that should allow running tests with any -j would be to copy missing dependencies from other makefiles to tests/Makefile. If I identified them correctly, this should make it work (for now):
````
lz4c unlz4 lz4cat: lz4
lz4: $(patsubst %.c,%.o,$(LZ4DIR)/*.c)
````
Additionally lz4c32 recipe first builds a 32-bit variant and then copies regular lz4 over it. Surely not what was supposed to happen.

**Comment by Cyan4973** [comment]
Latest update in `parallelMake` branch should fix the remaining issue regarding `make test -j6`.

> I had to also rm -f programs/lz4c before make install (where it was created again)

Could you describe your issue ?
Do you mean `make install` would not work after a `make test` ?

**Comment by itoffshore** [comment]
testing again I no longer need to `rm -f programs/lz4c` with the parallel patch

**Comment by Cyan4973** [comment]
Merged into `dev`
